Redundant pointerenter and mouseenter events fire for ancestors of removed child element underneath the cursor

lilyspiniolas opened the pull request:
#### 247b539986bd5cda1f20002a02c506cc344492fd
<pre>
Redundant pointerenter and mouseenter events fire for ancestors of removed child element underneath the cursor
<a href="https://bugs.webkit.org/show_bug.cgi?id=260550">https://bugs.webkit.org/show_bug.cgi?id=260550</a>
<a href="https://rdar.apple.com/114283232">rdar://114283232</a>

Reviewed by Abrar Rahman Protyasha.

In EventHandler::updateMouseEventTargetNode, track the last 32 elements underneath the
cursor and check if the current element underneath the mouse is an ancestor of the last
element underneath the mouse. If it is, skip firing mouseenter.

Tested by `pointerevents/pointerevent_after_target_removed.html?mouse`.

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_from_slot_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_lostpointercapture_for_disconnected_node-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_reappending_last_over_target-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_reappending_last_over_target.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=LEFT&amp;button=0&amp;buttons=1-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&amp;button=1&amp;buttons=4-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&amp;button=1&amp;buttons=4-expected.txt: Copied from LayoutTests/platform/glib/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&amp;button=1&amp;buttons=4-expected.txt.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=LEFT&amp;button=0&amp;buttons=1-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&amp;button=1&amp;buttons=4-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&amp;button=1&amp;buttons=4-expected.txt: Renamed from LayoutTests/platform/glib/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&amp;button=1&amp;buttons=4-expected.txt.
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateMouseEventTargetNode):
(WebCore::EventHandler::isElementAnAncestorOfLastElementUnderMouse const):
* Source/WebCore/page/EventHandler.h:

Canonical link: <a href="https://commits.webkit.org/300407@main">https://commits.webkit.org/300407@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f93c5b72bd124e1232f06bf794aee11b3604182

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122310 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42014 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32696 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128909 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74404 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b65539fa-d592-4aad-959c-ae24e28eeddc) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124186 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42731 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50608 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92978 "Found 1 new test failure: imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_reappending_last_over_target.tentative.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61788 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5d8821fb-f939-43e1-b325-9b28fc77901f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125262 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34084 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109520 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73637 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e3dd847d-4eee-402f-aae3-bb3f4f4ff93f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33085 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27680 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72384 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/103825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27887 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131637 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49250 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37474 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101527 "Found 1 new test failure: imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_reappending_last_over_target.tentative.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49624 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105740 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101398 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25747 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46763 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24891 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46012 "Build was cancelled. Recent messages:Printed configuration") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49107 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54847 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48576 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51926 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50257 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -3004,15 +3004,17 @@ void EventHandler::updateMouseEventTargetNode(const AtomString& eventType, Node*
+                elementsUnderMouse.append(element);
 
-            if (!leftElementsChain.isEmpty() && !enteredElementsChain.isEmpty() && leftElementsChain.last().ptr() == enteredElementsChain.last().ptr()) {
+            Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>, 32> enteredElementsChain = elementsUnderMouse;
```
I think you can drop the template args here
```suggestion
            Vector enteredElementsChain = elementsUnderMouse;
```

aprotyas commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -3031,9 +3033,16 @@ void EventHandler::updateMouseEventTargetNode(const AtomString& eventType, Node*
+                }
             }
+
+            m_ancestorsOfLastElementUnderMouse = elementsUnderMouse;
```
```suggestion
            m_ancestorsOfLastElementUnderMouse = WTFMove(elementsUnderMouse);
```

aprotyas commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -3066,6 +3075,19 @@ void EventHandler::clearElementUnderMouse()
+        return false;
+
+    for (WeakPtr weakElement : m_ancestorsOfLastElementUnderMouse) {
+        if (weakElement.get() == element)
```
This already exists:

```cpp
inline bool operator==(const WeakPtr<T, WeakPtrImpl, PtrTraits>& a, U* b)
```

so you can:
```suggestion
        if (weakElement == element)
```

lilyspiniolas committed:
Redundant pointerenter and mouseenter events fire for ancestors of removed child element underneath the cursor
https://bugs.webkit.org/show_bug.cgi?id=260550
rdar://114283232

Reviewed by Abrar Rahman Protyasha.

In EventHandler::updateMouseEventTargetNode, track the last 32 elements underneath the
cursor and check if the current element underneath the mouse is an ancestor of the last
element underneath the mouse. If it is, skip firing mouseenter.

Tested by `pointerevents/pointerevent_after_target_removed.html?mouse`.

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_from_slot_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_lostpointercapture_for_disconnected_node-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_reappending_last_over_target-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_reappending_last_over_target.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=LEFT&button=0&buttons=1-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&button=1&buttons=4-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&button=1&buttons=4-expected.txt: Copied from LayoutTests/platform/glib/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&button=1&buttons=4-expected.txt.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=LEFT&button=0&buttons=1-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&button=1&buttons=4-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&button=1&buttons=4-expected.txt: Renamed from LayoutTests/platform/glib/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&button=1&buttons=4-expected.txt.
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateMouseEventTargetNode):
(WebCore::EventHandler::isElementAnAncestorOfLastElementUnderMouse const):
* Source/WebCore/page/EventHandler.h:

Canonical link: https://commits.webkit.org/300407@main
